crates/holochain/src/conductor/conductor/app_connection_auth.rs

```diff
@@ -0,0 +1,243 @@
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[tokio::test]
+    async fn issue_and_use_single_use_token() {
+        let mut auth = AppAuthTokenStore::new();
+        let installed_app_id = "test_app".to_string();
+        let (token, _) = auth.issue_token(installed_app_id.clone(), 10, true);
+
+        let authenticated_for_app = auth.authenticate_token(token.clone(), None).unwrap();
+        assert_eq!(authenticated_for_app, installed_app_id);
+
+        let result = auth.authenticate_token(token.clone(), None);
+        assert!(result.is_err());
+    }
+
+    #[tokio::test]
+    async fn reuse_token() {
+        let mut auth = AppAuthTokenStore::new();
+        let installed_app_id = "test_app".to_string();
+        let (token, _) = auth.issue_token(installed_app_id.clone(), 10, false);
+
+        let authenticated_for_app = auth.authenticate_token(token.clone(), None).unwrap();
+        assert_eq!(authenticated_for_app, installed_app_id);
+
+        let authenticated_for_app = auth.authenticate_token(token.clone(), None).unwrap();
+        assert_eq!(authenticated_for_app, installed_app_id);
+    }
+
+    #[tokio::test]
+    async fn attempt_with_token_that_does_not_exist() {
+        let mut auth = AppAuthTokenStore::new();
+        let result = auth.authenticate_token(vec![0; 16], None);
+        assert!(result.is_err());
+    }
+
+    #[tokio::test]
+    async fn use_token_with_app_restriction() {
+        let mut auth = AppAuthTokenStore::new();
+        let installed_app_id = "test_app".to_string();
+        let (token, _) = auth.issue_token(installed_app_id.clone(), 1, true);
+
+        let result = auth.authenticate_token(token.clone(), Some(installed_app_id.clone()));
+        assert_eq!(result.unwrap(), installed_app_id);
+    }
+
+    #[tokio::test]
+    async fn use_token_with_app_restriction_mismatch() {
+        let mut auth = AppAuthTokenStore::new();
+        let installed_app_id = "test_app".to_string();
+        let (token, _) = auth.issue_token(installed_app_id.clone(), 1, true);
+
+        let other_app_id = "other_app".to_string();
+        let result = auth.authenticate_token(token.clone(), Some(other_app_id));
+        assert!(result.is_err());
+
+        // Token was invalidated by the use in a failed attempt
+        let result = auth.authenticate_token(token.clone(), Some(installed_app_id.clone()));
+        assert!(result.is_err());
+    }
+
+    #[tokio::test]
+    async fn use_token_with_app_restriction_mismatch_multi_use() {
+        let mut auth = AppAuthTokenStore::new();
+        let installed_app_id = "test_app".to_string();
+        let (token, _) = auth.issue_token(installed_app_id.clone(), 1, false);
+
+        let other_app_id = "other_app".to_string();
+        let result = auth.authenticate_token(token.clone(), Some(other_app_id));
+        assert!(result.is_err());
+
+        // Token was retained through the failed attempt. Another reason that single use tokens are safer!
```

Comment:
okay, it seems like you want to give the option of multi-use tokens being validate despite incorrect app id

Reply:
Yes I do, justified above and I've expanded on this comment